rust/extractor/src/crate_graph.rs

```diff
@@ -0,0 +1,1324 @@
+    // Each crate is defined by the `FileId` of its root module, the set of enabled
+    // `cfg` flags and the set of dependencies.
+
+    // First compute a hash map with for each crate ID, the path to its root module and a hash of all
+    // its `cfg` flags and dependencies hashes. Iterate in topological order to ensure hashes of dependencies
+    // are present in the map.
+    let mut crate_id_map = HashMap::<CrateId, (PathBuf, u64)>::new();
+    for krate_id in crate_graph.crates_in_topological_order() {
+        let krate = &crate_graph[krate_id];
+        let root_module_file: &VfsPath = vfs.file_path(krate.root_file_id);
+        if let Some(root_module_file) = root_module_file
+            .as_path()
+            .map(|p| std::fs::canonicalize(p).unwrap_or(p.into()))
+        {
+            let mut hasher = std::hash::DefaultHasher::new();
+            krate
+                .cfg_options
+                .as_ref()
+                .into_iter()
+                .sorted_by(cmp_flag)
+                .for_each(|x| format!("{x}").hash(&mut hasher));
+
+            krate
+                .dependencies
+                .iter()
+                .flat_map(|d| crate_id_map.get(&d.crate_id))
+                .sorted()
+                .for_each(|x| x.hash(&mut hasher));
+            let hash = hasher.finish();
+            crate_id_map.insert(krate_id, (root_module_file, hash));
+        }
+    }
+    // Extract each crate
+    for krate_id in crate_graph.iter() {
+        if let Some((root_module_file, hash)) = crate_id_map.get(&krate_id) {
+            let path = root_module_file.join(format!("{hash:0>16x}"));
+            let mut trap = trap_provider.create("crates", path.as_path());
+            // If the trap file already exists, then skip extraction because we have already extracted
+            // this crate with the same `cfg` flags and dependencies.
+            // FIXME: this may need to b improved in case the implemenation becomes multi-threaded
+            if trap.path.exists() {
+                continue;
+            }
+            let krate = &crate_graph[krate_id];
+            let module = emit_module(
+                &crate_graph,
+                db,
+                db.crate_def_map(krate_id).as_ref(),
+                "crate",
+                DefMap::ROOT,
+                &mut trap,
+            );
+            let file_label = trap.emit_file(root_module_file).as_untyped();
+            let location_label = codeql_extractor::extractor::location_label(
+                &mut trap.writer,
+                Location {
+                    file_label,
+                    start_line: 0,
+                    start_column: 0,
+                    end_line: 0,
+                    end_column: 0,
+                },
+            );
+            trap.writer.add_tuple(
+                "locatable_locations",
+                vec![module.into(), location_label.into()],
+            );
+
```

Comment:
Nit: I'd say this block may be moved to the `emit_module` function, as I'd say it's part of the "module" emission

Reply:
ah, I had missed that! Maybe then we could rename `module` to `root_module` for extra clarity? Also, it may make sense to add a `fn emit_file_only_location(&mut self, file_label: Label<generated::File>, entity_label: Label<E>)` to `TrapFile` doing this whole block (and maybe internally reusing some of the code in `emit_location` and `emit_location_label`, so that here we could just write
```suggestion
            let file_label = trap.emit_file(root_module_file);
            trap.emit_file_only_location(root_module, file_label);
```

Reply:
but that's ok also left for follow-up changes.